File: frontend/src/stores/maintenance.ts
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'
import { errorMessage } from '@/i18n/errors'
import { Events } from '@/lib/events'
import { parseBackendError, type BackendError } from '@/lib/backend-error'
import { eventsOn, quitApp } from '@/lib/wails-runtime'
import type { BackupResult, OptimizeResult } from '@/types/maintenance'

export const useMaintenanceStore = defineStore('maintenance', () => {
  const running = ref(false)
  const confirmClose = ref(false)
  const closingApp = ref(false)
  const lastError = ref<BackendError | null>(null)
  const optimizeResult = ref<OptimizeResult | null>(null)
  const backupResult = ref<BackupResult | null>(null)

  const busy = computed(() => running.value)

  let listening = false

  function listen() {
    if (listening) {
      return
    }
    listening = true
    eventsOn(Events.MaintenanceCloseRequested, () => {
      confirmClose.value = true
    })
  }

  async function refreshRunning() {
    running.value = await window.go.handlers.App.DatabaseMaintenanceRunning()
  }

  async function optimize(): Promise<OptimizeResult | null> {
    lastError.value = null
    optimizeResult.value = null
    running.value = true
    try {
      const result = await window.go.handlers.App.OptimizeDatabase()
      optimizeResult.value = result
      backupResult.value = null
      return result
    } catch (err) {
      lastError.value = parseBackendError(err)
      throw err
    } finally {
      running.value = false
    }
  }

  async function backup(): Promise<BackupResult | null> {
    lastError.value = null
    backupResult.value = null
    running.value = true
    try {
      const result = await window.go.handlers.App.CreateCatalogBackup()
      if (result?.path) {
        backupResult.value = result
      }
      return result
    } catch (err) {
      lastError.value = parseBackendError(err)
      throw err
    } finally {
      running.value = false
    }
  }

  function stayInApp() {
    confirmClose.value = false
    if (!closingApp.value) {
      void window.go.handlers.App.DismissWindowClose()
    }
  }

  function leaveApp() {
    closingApp.value = true
    confirmClose.value = false
    quitApp()
  }

  function errorText() {
    if (!lastError.value) {
      return ''
    }
    return errorMessage(lastError.value.code, lastError.value.params)
  }

  return {
    running,
    busy,
    confirmClose,
    lastError,
    optimizeResult,
    backupResult,
    listen,
    refreshRunning,
    optimize,
    backup,
    stayInApp,
    leaveApp,
    errorText,
  }
})

File: internal/services/app/maintenance.go
package app

import (
	"context"

	"github.com/alexbelweb/flibustahub/internal/apperr"
	"github.com/alexbelweb/flibustahub/internal/services/maintenance"
)

func (s *Service) OptimizeDatabase(ctx context.Context) (maintenance.OptimizeResult, error) {
	if s == nil || s.maint == nil {
		return maintenance.OptimizeResult{}, apperr.New(apperr.CodeDBOpenFailed, nil)
	}
	return s.maint.Optimize(ctx)
}

func (s *Service) CreateCatalogBackup(ctx context.Context) (maintenance.BackupResult, error) {
	if s == nil || s.maint == nil {
		return maintenance.BackupResult{}, apperr.New(apperr.CodeDBOpenFailed, nil)
	}
	return s.maint.Backup(ctx)
}

func (s *Service) DatabaseMaintenanceRunning() bool {
	if s == nil || s.maint == nil {
		return false
	}
	return s.maint.Running()
}

func (s *Service) CancelMaintenance() {
	if s == nil || s.maint == nil {
		return
	}
	s.maint.Cancel()
}
